contents/javascript/Javascript-framework-React.md

```diff
@@ -0,0 +1,152 @@
+
+To see how this works, let's build the same web application using React instead:
+
+<box type="info">
+    The sandbox below is editable. You can play around with the application data (or <code>state</code>) and see how the UI automatically gets updated. 
+</box>
+
+<iframe src="https://codesandbox.io/embed/p564px0j47?fontsize=12" style="width:100%; height:400px; border:0; border-radius: 4px; overflow:hidden;" sandbox="allow-modals allow-forms allow-popups allow-scripts allow-same-origin"></iframe>
+
+As you can see above, the declarative approach is useful because it abstracts the complexity of interacting with the actual DOM elements and <tooltip content="React provides its own event system so that events can work in the same way across different browsers.">events</tooltip>. With React, you just need to specify *what* you want to show in the UI when the application data changes.
+
+#### Benefit: Reusability
+
+Suppose you have a web application with the following markup in its UI:
+
+```html
+<h3>This is line 1.</h3>
+<h3>This is line 2.</h3>
+<h3>This is line 3.</h3>
+```
+
+The only difference in the three headings above is the number at the end of each line. Wouldn't it be nice if we could reuse a single line of markup for each heading and just pass in a different number each time?
+
+We can do this with React! Let's see how:
+
+<iframe src="https://codesandbox.io/embed/884m88wr90?fontsize=12" style="width:100%; height:400px; border:0; border-radius: 4px; overflow:hidden;" sandbox="allow-modals allow-forms allow-popups allow-scripts allow-same-origin"></iframe>
+
+<box type="info">
+    Notice the HTML-like syntax within the JavaScript method <code>render()</code>. This is <a href="https://reactjs.org/docs/introducing-jsx.html">JSX</a>, and it is a JavaScript syntax extension developed by the React team. React <a href="https://reactjs.org/docs/introducing-jsx.html#why-jsx">recommends</a> developers to use JSX, but it is optional to use it.
+</box>
+
+In the example above, we reused the markup for each heading using an `ExampleComponent`. A React **component** is just a JavaScript function that accepts arbitrary input (called `props`) and returns elements describing what should appear on the screen. 
+
+Components can be especially useful when your UI is complex. With components, you can not only *reuse* code, but also *divide* your UI into encapsulated units that manage their own data and logic. 
+
+For example, if you have a todo list application, it could have a `Form` component to handle adding new items and a `TodoList` component to display all the items. The `TodoList` component could further be composed of multiple `TodoListItem` components, one for each item in the todo list. 
+
+![Building a Todo List application using React](javascript-framework-react-imgs/ReactComponentsExample.png "Building a Todo List application using React")
```

Comment:
Did you make this yourself? If so, nice work :)
If not, or if it was modified, provide a link to the source please

Reply:
Modified `https://i.cloudup.com/DCcBG58qYv.png`, I'll put that in.